### PR TITLE
proposed fix to issue #492

### PR DESCRIPTION
**Pull request by ShijunS**
The proposal is to queue a task to fire the corresponding events to avoid risk of reentrancy (#492).

### CODE REVIEW

File: webrtc.html [hunk 1]
```diff
@@ -720,7 +720,7 @@
                   </li>
 
                   <li>
-                    <p>Fire a simple event named <code><a href=
+                    <p>Queue a task to fire a simple event named <code><a href=
```
Comment (jan-ivar) [inline webrtc.html]: We are already in a queued task here, so reentering that task is impossible.

Nor is there any risk of garbage collection, since the peer connection is held alive by the stack.

This change alters the order in which a signalingstatechange event-handler and a promise onRejected function are called. Is that desirable?

File: webrtc.html [hunk 2]
```diff
@@ -4685,7 +4685,7 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
         <li>
           <p>If the <a href="#dfn-underlying-data-transport">transport</a> was
           closed <dfn id="data-transport-closed-error">with an error</dfn>,
-          fire an NetworkError event at <var>channel</var>.</p>
+          queue a task to fire an NetworkError event at <var>channel</var>.</p>
```
Comment (jan-ivar) [inline webrtc.html]: Same here.

File: webrtc.html [hunk 3]
```diff
@@ -866,7 +866,7 @@
 
                   <li>
                     <p>If <var>connection</var>'s <a>signaling state</a>
-                    changed above, fire a simple event named <code><a href=
+                    changed above, queue a task to fire a simple event named <code><a href=
```
Comment (jan-ivar) [inline webrtc.html]: This one makes sense I think, though I observe that elsewhere (like for ICE gathering state directly below) we tend to queue a task to set the state *and* fire the event, though here it may matter more that the state is set early? Should we worry that the state may have changed again by the time the event fires? I'm unsure.

Comment (alvestrand) [inline webrtc.html]: We have had issues with the state being overwritten by a second event after the task was queued but before the event was fired, so yes, I think we should set the state within the queued task if we queue it (otherwise, the user observes a surprising state transition).

Comment (stefhak) [inline webrtc.html]: @alvestrand are you arguing we should leave things as they are (i.e. we queue a task that potentially changes the state, and if it does it fires the event rather than queuing another task to fire the event)?

File: webrtc.html (hunk 2 above)
Comment (stefhak) [inline webrtc.html]: (maybe this is what you're saying @jan-ivar ) this change would mean we change the order of the NetworkError and close events.

File: webrtc.html (hunk 1 above)
Comment (adam-be) [inline webrtc.html]: We want the state update to happen synchronously as the event that tells the script about it, so we don't want to queue an other task here. @jan-ivar, does it matter if we fire the event first or reject the promise? I.e. change the order of step 6 and 7 in this algorithm.

File: webrtc.html (hunk 3 above)
Comment (alvestrand) [inline webrtc.html]: Looking back, I think it's right as it is (no change). The event should fire after the state change, but before any possible later state change.
(It took me a while to realize that "fire an event" means "call the event handler NOW", not "put an event in a queue and call the event handler later").

Comment (adam-be) [inline webrtc.html]: Note that all operations like "update connection's ICE gathering/connection state" are queued in the steps that are liked to.

The intention here is to have the state updated and receivers dispatched for all new incoming media before the promise "then-handler" is run. So signalingstatechange needs to be fired synchronously as the state is updated. The question is we need to the synchronous work later in the steps.

File: webrtc.html (hunk 2 above)
Comment (adam-be) [inline webrtc.html]: This one is a bit strange. Is the "NetworkError event" even defined somewhere? I think we could do with one event here.

File: webrtc.html (hunk 1 above)
Comment (jan-ivar) [inline webrtc.html]: @adam-be the difference would only be observable in a JS debugger, because a script can only learn of a rejection through functions passed to `.then` or `.catch`, which are guaranteed to always execute on a microtask.

File: webrtc.html (hunk 3 above)
Comment (jan-ivar) [inline webrtc.html]: @adam-be But as you just mentioned, "update connection's ICE gathering/connection state" are queued. That means they already happen *after* functions passed to `.then` execute, because promises operate on the microtask queue, whereas "queue a task" refers to the main event loop.

Comment (jan-ivar) [inline webrtc.html]: Also, of the three changes in this patch, I think @ShijunS has a point on this one that it's perhaps unwise to dictate substantive steps involving non-stack objects after requiring the UA to synchronously execute content script code, as it's a window to potentially trip up assumptions in the remaining steps in some implementations. For instance, an implementation that didn't properly root the peer connection object on the stack in this queued task, might not take kindly to content script removing its last peer connection reference. I believe that's what he's referring to as reentrancy.

Comment (jan-ivar) [inline webrtc.html]: Although I just spoke with our DOM team and they're saying that "In practice, web algorithms are treated as closing over everything they might care about that they don't get from somewhere explicitly".

File: webrtc.html (hunk 1 above)
Comment (ShijunS) [inline webrtc.html]: It looks like there is no objection to @jan-ivar's comment.   Meanwhile, the Mozilla team has confirmed offline that there is no concern from browser implementation perspective.  I will remove this specific change from the proposal.

File: webrtc.html (hunk 2 above)
Comment (ShijunS) [inline webrtc.html]: Based on comment from @Jan-Ivar and feedback from the Mozilla team, I will remove the specific change from the proposal.  @adam-be, it might be better to open a new issue to address the question on the NetworkError event.

**Comment by adam-be** [comment]
> Based on comment from @Jan-Ivar and feedback from the Mozilla team, I will remove the specific change from the proposal. @adam-be, it might be better to open a new issue to address the question on the NetworkError event.

I got this comment from @ShijunS in my email inbox, but I can't find it on github so I'm quoting it. @ShijunS, please do create a new issue on the RTCDataChanel NetworkError event.

**Comment by ShijunS** [comment]
@adam-be, issue #526 is opened to track the question on NetworkError event.

**Comment by adam-be** [comment]
Thank you @ShijunS

File: webrtc.html (hunk 3 above)
Comment (adam-be) [inline webrtc.html]: If we want to queue a new task here I think we want to do the state update in that task as well. But I didn't really understand from @jan-ivar if it was necessary. What do you say @jan-ivar?

**Comment by alvestrand** [comment]
@jan-ivar this was the PR I wanted your comment on - the proposed 1-line change would queue a task from within a task that's already been queued. You tend to have thought about this kind of thing.

**Comment by jan-ivar** [comment]
I don't know if a change is necessary. @ShijunS seemed to feel strongly about it though. It is certainly true that, after having made a synchronous call to content JS, one must either re-check that one still has everything one needs, or make sure to follow a model where things are well-rooted (reference-counted) at the start of the task, such that reference-counts wont go to zero and GC'ed objects could be encountered. To me, that would only seem to be problematic in poorly written implementations, e.g. C++ code that stores a raw pointer on the stack before the synchronous call to JS, and relies on using it again after.

It would be easier if we didn't make this change, because even though I agree with @adam-be that it'd make sense to want to do the state update in that new task as well, doing so would update the state too late to be observable in the success callback, because a "queued task" happens on the next JS event loop, which is later than `.then` success callbacks which happen on the micro-task queue. (The micro-task queue is completely emptied at the end of the current run-to-completion in most browsers).

**Comment by jan-ivar** [comment]
TL;DR: If I go by our DOM team, then no change is necessary (they say "in practice, web algorithms are treated as closing over everything they might care about that they don't get from somewhere explicitly").

**Comment by ShijunS** [comment]
There are events queued in steps 6 and 7 in the same algorithm.  I expect it will be safer to queue the event at step 5 as well.  All the events in the specific algorithm will at least be queued in the same place so stay in the same order.  @Alvestrand, it seems some browser implementation might have the problem described by @Jan-Ivar.  Meanwhile, the Firefox folks have confirmed there is no problem with the Firefox implementation.  So, I don't feel as strongly as previously about the potential risk.  It can be considered as an implementation issue, unless other browser vendor has different preference.

**Comment by adam-be** [comment]
I believe the current text reflects the groups desire to have the signalling state updated by the time the promise resolved (as @jan-ivar talks about above). Should we close this for now and reopen if people have a hard time implementing it as written?

**Comment by jan-ivar** [comment]
wfm.

**Comment by alvestrand** [comment]
Conclusion is that we should close this without merging. Branched-off discussions have been moved to other bugs.
